volatility3/framework/symbols/linux/utilities/modules.py

```diff
@@ -210,77 +225,173 @@ def get_kernel_module_info(
+        An exception is thrown if an empty source list is given or a list containing an invalid source
+        """
+        if not caller_wanted_sources:
+            raise ValueError("`caller_wanted_sources` must have at least one source.")
+
+        if (
+            len(caller_wanted_sources) == 1
+            and caller_wanted_sources[0] == Modules.source_hidden_identifier
+        ):
+            raise ValueError(
+                f"{Modules.source_hidden_identifier} cannot be the only source or there is nothing to compare against."
+            )
+
+        wanted_sources = []
+
+        for source in caller_wanted_sources:
+            if source not in Modules.all_sources_identifier:
+                raise ValueError(
+                    f"Invalid source sent through `caller_wanted_sources`: {source}"
+                )
+
+            wanted_sources.append(source)
+
+        return wanted_sources
+
     @classmethod
     def run_modules_scanners(
         cls,
         context: interfaces.context.ContextInterface,
-        kernel_name: str,
-        run_hidden_modules: bool = True,
+        kernel_module_name: str,
+        caller_wanted_sources: List[str],
         flatten: bool = True,
     ) -> Dict[str, List[ModuleInfo]]:
         """Run module scanning plugins and aggregate the results. It is designed
         to not operate any inter-plugin results triage.
 
+        Rules for `caller_wanted_sources`:
+
+            If `Modules.all_sources_identifier` is specified then every source will be populated
+
+            If `Modules.source_hidden_identifier` is in the list, then at least one other sources must be
+            specified so a comparison will be populated
+
+            If empty or an invalid source is specified then a ValueError is thrown
+
         Args:
-            run_hidden_modules: specify if the hidden_modules plugin should be run
+            called_wanted_sources: The list of sources to gather modules.
+            flatten: Whether to de-duplicate modules across sources
         Returns:
             Dictionary mapping each plugin to its corresponding result
         """
 
-        kernel = context.modules[kernel_name]
+        module_gatherers = {
+            Modules.source_kernel_identifier: cls.get_kernel_module_info,
+            Modules.source_lsmod_identifier: cls._get_list_modules,
+            Modules.source_sysfs_identifier: cls._get_sysfs_modules,
+        }
+
+        kernel = context.modules[kernel_module_name]
 
         address_mask = context.layers[kernel.layer_name].address_mask
 
+        wanted_sources = Modules._validated_sources(caller_wanted_sources)
+
         run_results = {}
 
-        # the kernel module boundaries
-        run_results["kernel"] = [cls.get_kernel_module_info(context, kernel_name)]
+        run_hidden_modules = False
 
-        # lsmod
-        run_results["lsmod"] = []
+        # Special case hidden modules since it gathers modules on its own
```

Comment:
Special casing this makes it somewhat difficult for other gatherers to be written that could work the same...  5:S  Would it be better to make an *all* modules gatherer and run a process over the result to identify hidden ones (basically, do the filtering after)?  That would then allow all gatherers to operate in the same way, and the results would be more consistent and all the special casing goes away (apart from a step at the end to filter one type from the other)?